### ISSUE TITLE
Use RawHttp to implement a proxy

### ISSUE THREAD

**Issue by apflieger** [renatoathaydes/rawhttp#34]
Hi,

I'm trying to use RawHttp to implement a simple http proxy but I encounter some problems. Here's what I try to do

```java
    public static void main(String[] args) {
        RawHttpServer proxyServer = new TcpRawHttpServer(80);
        TcpRawHttpClient client = new TcpRawHttpClient();
        proxyServer.start(clientRequest -> {
            try {
                // Rewrite the request to point the end server
                clientRequest = clientRequest.withRequestLine(clientRequest.getStartLine().withHost("localhost:50005"));
                // Forward the request to the end server
                RawHttpResponse<Void> response = client.send(clientRequest);
                // Return the response as it is
                return Optional.of(response);
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
    }
```

It seems to work great for GET requests but when I try a POST, things breaks. It appears that initial client gets a response with an empty body.
Here's a curl example on a /partenaires route that my end server can handle.

```
curl 'http://localhost/partenaires'   --data-raw 'email=iiuzhef%40izuef.com&school-name=ezfzef&postal-code=ozief&phone=fze'   --compressed -v
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 80 (#0)
> POST /partenaires HTTP/1.1
> Host: localhost
> User-Agent: curl/7.64.1
> Accept: */*
> Accept-Encoding: deflate, gzip
> Content-Length: 72
> Content-Type: application/x-www-form-urlencoded
>
* upload completely sent off: 72 out of 72 bytes
* Empty reply from server
* Connection #0 to host localhost left intact
curl: (52) Empty reply from server
* Closing connection 0
```

The POST request is forwarded and processed properly on the end server but the response is not sent to the original client (curl).
Is there a static state that makes `RawHttpServer` and `TcpRawHttpClient` not working together? Is there a special treatment of POST requests that could cause that? (again, it's working for GET requests)

**Comment by renatoathaydes** [comment]
Apart from the problem where the port was not being replaced when you called `withRequestLine` (which I fixed with the commit above), this has proven to be an extremely annoying bug.
RawHTTP has no global, mutable state. However, when I run your code and we try to send a HTTP request from within a HTTP Server Router, it seems that the JVM simply stops executing the code.

No Exception or Error is thrown. It just stops.

I've never seen this in over 10 years working with Java.

When the server stops executing, the client just receives no data, so the application fails with `InvalidHttpResponse{message='No content', lineNumber=0}`.

I will continue investigating this issue, but so far, I'm sorry but this looks like the JVM doing something very strange, not like a bug in RawHTTP itself (every time I said this before, I regretted!! :) but at least for now, that's what it's looking like).

**Comment by renatoathaydes** [comment]
Even with this code being used to proxy, the error still happens:

```java
Socket s = new Socket(clientRequest.getUri().getHost(), clientRequest.getUri().getPort());
clientRequest.writeTo(s.getOutputStream());
EagerHttpResponse<Void> response = new RawHttp().parseResponse(s.getInputStream()).eagerly();
System.out.println("Response from target: " + response.eagerly());
```

That's why I can rule out a bug in the RawHTTP Client. But because when I debug, the debugger seems to simply "die" at the line where we try to read from the socket's input stream, I cannot understand what's happening.

**Comment by apflieger** [comment]
Yeah I debugged it for like an entire day, I couldn't understand what happens.
I didn't see the code stopping, what I found instead was that the connexion to the initial client closes randomly during `client.send(clientRequest)`, the response of this call was correct and returned by the router correctly but since the connexion was already closed, the response is not sent back to the initial client.

**Comment by renatoathaydes** [comment]
If you look at the code for [`TcpRawHttpServer`](https://github.com/renatoathaydes/rawhttp/blob/master/rawhttp-core/src/main/java/rawhttp/core/server/TcpRawHttpServer.java#L295) you will see that we should always return a response, even if you route throws an Exception (even if I catch `Throwable` in that block, it never logs anything, but the next line doesn't execute : hence why I say the code just "stops").

It should be impossible for the client to just get no data at all. I am still struggling here to find an explanation.

**Comment by renatoathaydes** [comment]
Alright, I think I fixed the problem. This was very tricky because RawHTTP "intentionally" does not report socket errors caused by a closed socket, because these kinds of errors are common (sockets are closed every time a client's connection expires after too long idle, and sometimes even on each request if the client has no connection pool)...

Well, the problem was that the server was trying to send a response but the client's socket was already closed.

The client was closing its own socket after a POST request because it had a body, and I had made a mistake in the implementation of `HttpMessage#witeTo` in which I closed the `BodyReader`, and in case the body reader was still attached to a "live" connection, as in this case, the connection itself would be dropped.

At the same time, the client was waiting for a response on the same socket, so it ended up getting no data, which caused the `InvalidHttpResponse` error.

The Server would try to write the response but it got a `SocketException` (due to the client having closed the socket on its side) but as this is commonly because the client decided to not wait for the response, which is normal behaviour, the server was not reporting that error!!

I pushed a sample showing how proxying now works fine (I linked the commits to this ticket so you can see for yourself, @apflieger ).

I will make a release either later today or tomorrow as there's already quite a few good fixes.

**Comment by renatoathaydes** [comment]
Released the fix in version 2.4.0
